### PR TITLE
Constprop

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
perform constant propagation for add of const.

(c1 + c2) -> c3  for float and int

also perform normalization:

(c+x) -> (x + c)
(x+c1) + c2) -> x + (c1 + c2)

added tests as well; not 100% sure the pass is added on the normal onnx-mlir, maybe someone could tell me if it runs (or how to test what runs)

I also added new operation generators for all ops with broadcast, to generate the actual output size instead of an ?xtype.

In the process, also edited small typo in the (excellent) user page on how to recompile the ONNX doc to get the td files.

### CODE REVIEW

**Comment by tungld** [comment]
Be careful when building an op with ranked type but unknown dimensions like this `"onnx.Mul"(%arg0, %arg0) : (tensor<?x?x?xf32>, tensor<?x?x?xf32>) -> tensor<?x?x?xf32>`, which does not trigger shape inference for `Mul` and output type `tensor<?x?x?xf32>` remains unchanged after shape inference.

**Comment by AlexandreEichenberger** [comment]
@tungld Are you suggesting that I add a constraint for the "constant + constant" operation simplification? I assume that since they are constant, all of their dimensions must be known, but I can add a check if that is what you are asking for.

Or are you saying that when we have a (x+c1)+c2 I have to worry about broadcast rules disabling the x + (c1+c2) ? that is interesting, we would have to see if associativity is impacted by broadcast.

Please clarify your comment as I want to benefit from your insight, thanks.

**Comment by tungld** [comment]
Sorry for the confusion.

My comment is about custom builders. I saw you added some custom builders for broadcasting ops, where result types are ranked and potentially with unknown dimensions, for example the lit tests for decomposition you showed:
```mlir
// CHECK-NEXT: [[MUL:%.+]] =  "onnx.Mul"(%arg0, %arg0) : (tensor<?x?x?xf32>, tensor<?x?x?xf32>) -> tensor<?x?x?xf32>
// CHECK-NEXT: [[REDUCE_SUM:%.+]] = "onnx.ReduceSum"([[MUL]]) {axes = [1], keepdims = 0 : i64} : (tensor<?x?x?xf32>) -> tensor<*xf32>
```

In this case, when we run shape inference, let's say,`%arg0` gets fixed shape `tensor<10x10x10xf32>`. When shape inference comes to `Mul`, since `Mul`'s result type is a ranked type, shape inference for `Mul` is not triggered. Thus, result type of `Mul` still has unknown dimensions.

This happens because 
- Decompose pass is done before shape inference pass, and
- We currently trigger shape inference if the result type is not None and unranked: https://github.com/onnx/onnx-mlir/blob/master/src/Transform/ONNX/ShapeInferencePass.cpp#L135.

**Comment by chentong319** [comment]
1. This patch touched an important topic for onnx-mlir. It seems to me more of constant folding and canonicalization, other than constant propagation.
2. Any more folding and canonicalization oppotunities in addition to AddOp? Multiple with zero or one? Logical operation And, Not and Or?
3. Does this transformation require shape of output tensor known (shapedTensor), namely, shape inference pass has been applied? It seems unnecessary to me. When constant folding is performed, the shape inference is a byproduct. Then this pass can be invoked before shape inference.
4. The add broadcast is the complicated part. Does onnx-mlir have a function to handle broadcast for all the element-wise operation? I assume that the broadcast has already been implemented somewhere else and did not go into detail of your implementation.

**Comment by AlexandreEichenberger** [comment]
> Sorry for the confusion.
> 
> My comment is about custom builders. I saw you added some custom builders for broadcasting ops, where result types are ranked and potentially with unknown dimensions, for example the lit tests for decomposition you showed:
> ...
> In this case, when we run shape inference, let's say,`%arg0` gets fixed shape `tensor<10x10x10xf32>`. When shape inference comes to `Mul`, since `Mul`'s result type is a ranked type, shape inference for `Mul` is not triggered. Thus, result type of `Mul` still has unknown dimensions.
> 
> This happens because
> 
> * Decompose pass is done before shape inference pass, and
> * We currently trigger shape inference if the result type is not None and unranked: https://github.com/onnx/onnx-mlir/blob/master/src/Transform/ONNX/ShapeInferencePass.cpp#L135.

The reason I had to add the custom builder is that with the Declarative Rewrite Rules generate a builder that was not present. The problem with constant propagation is that it is nice to have the  shape inference results maintained during the transformation. Maybe I can add a test in the new builder that check if all dimensions are known, in which case I return the broadcast type; and if they are not known, then I return the previous unknown value

**Comment by AlexandreEichenberger** [comment]
> * This patch touched an important topic for onnx-mlir. It seems to me more of constant folding and canonicalization, other than constant propagation.

thanks

> * Any more folding and canonicalization oppotunities in addition to AddOp? Multiple with zero or one? Logical operation And, Not and Or?

of course, use templates, to it will be very easy to add any other ops that use broadcast; other ops are much easier to handle. I plan to add all the transform listed here:

```
Removes trivial ops, e.g. identity Reshape, Transpose of 1-d tensors, Slice(x) = x, etc.
● Rewrites that enable further constant folding, e.g.
○ Constant propagation through Enter
○ Switch(pred=x, value=x) => propagate False through port0, True through port1
○ Partial constant propagation through IdentityN
● Arithmetic rewrites that rely on known shapes or inputs, e.g.
○ Constant push-down:
■ Add(c1, Add(x, c2)) => Add(x, c1 + c2)
■ ConvND(c1 * x, c2) => ConvND(x, c1 * c2)
○ Partial constfold:
■ AddN(c1, x, c2, y) => AddN(c1 + c2, x, y),
■ Concat([x, c1, c2, y]) = Concat([x, Concat([c1, c2]), y)
○ Operations with neutral & absorbing elements:
■ x * Ones(s) => Identity(x), if shape(x) == output_shape
■ x * Ones(s) => BroadcastTo(x, Shape(s)), if shape(s) == output_shape
■ Same for x + Zeros(s) , x / Ones(s), x * Zeros(s) etc.
■ Zeros(s) - y => Neg(y), if shape(y) == output_shape
■ Ones(s) / y => Recip(y) if shape(y) == output_shape
 ```

> * Does this transformation require shape of output tensor known (shapedTensor), namely, shape inference pass has been applied? It seems unnecessary to me. When constant folding is performed, the shape inference is a byproduct. Then this pass can be invoked before shape inference.

Maybe, but I need to apply the DRR (declarative rewrite rules), and I don't think that shape inference fits in that pattern... but if this was not the case, I am happy to do it.

As a side comment, when dealing with constant, all the dimensions must be known at compile time... otherwise its not a constant array.

> * The add broadcast is the complicated part. Does onnx-mlir have a function to handle broadcast for all the element-wise operation? I assume that the broadcast has already been implemented somewhere else and did not go into detail of your implementation.

onnx-mlir has lowering using broadcast rules... here I need to compute the values. I could not find an easy fit. In lowering, you use kernel iterate to generate loops, here I need to execute the loops. 

If you find a way to reuse the code, I am happy to look into it.


Thanks for your insight, Tong.

**Comment by chentong319** [comment]
Yes, all the attributes should have known shape. I thought you checked the shape of output of Op in the code. But Only the shape is assumed known in your test case. I mixed the two.

File: src/MainUtils.cpp [hunk 1]
```diff
@@ -88,6 +88,7 @@ void registerDialects() {
 
 void addONNXToMLIRPasses(mlir::PassManager &pm) {
   pm.addPass(mlir::createDecomposeONNXToONNXPass());
+  pm.addPass(mlir::createConstPropONNXToONNXPass());
```
Comment (tjingrant) [inline src/MainUtils.cpp:91]: Might read better without the "ONNXToONNX" part.

File: src/Transform/ONNX/ONNXConstProp.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Pass/Passes.hpp"
+
+using namespace mlir;
+
+namespace {
+
+////////////////////////////////////////////////////////////////////////////////
```
Comment (tjingrant) [inline src/Transform/ONNX/ONNXConstProp.cpp:28]: Comment style is off w.r.t. other MLIR function/class level comments. I believe it's "=" rather than "/".

**Comment by AlexandreEichenberger** [comment]
@tungld 

To address your concerns, I now generate the broadcast type only when type has a static shape. Change is in the python script, but here is the output after python gen and table gen. Otherwise, its the same unranked tensor as before. I reverted the change in the lit tests too.

```C++
void ONNXAddOp::build(OpBuilder &builder, OperationState &state, Value A, Value B) {
      auto lhsTy = A.getType().cast<RankedTensorType>();
      auto rhsTy = B.getType().cast<RankedTensorType>();
      auto elementType = getBroadcastedType(lhsTy, rhsTy);
      auto shapedType = elementType.dyn_cast_or_null<ShapedType>();
      if (!shapedType || !shapedType.hasStaticShape()) {
            elementType = A.getType().cast<TensorType>().getElementType();
            elementType = UnrankedTensorType::get(elementType);      }
      build(builder, state, elementType, A, B);
    
}
```

**Comment by AlexandreEichenberger** [comment]
@chentong319 

To address your concerns, I added the subtract, unary support to handle the NEG, and multiply. I will leave the handling of operations with known constants (like Ones, Zeros) for another PR, as this one is already pretty big. Think the concat of constant is probably important as well.

Note that adding new ops is very simple, will add some comment in the code to this effect.

File: src/MainUtils.cpp (hunk 1 above)
Comment (AlexandreEichenberger) [reply src/MainUtils.cpp:91]: Do you want me to remove the ONNXToONNX for both? or change it to O2O?

File: src/Transform/ONNX/ONNXConstProp.cpp (hunk 2 above)
Comment (AlexandreEichenberger) [reply src/Transform/ONNX/ONNXConstProp.cpp]: changed as requested

**Comment by AlexandreEichenberger** [comment]
added instructions in the files on how to add new ops.

File: src/MainUtils.cpp (hunk 1 above)
Comment (chentong319) [reply src/MainUtils.cpp:91]: Consider "ConstantFoldingPass"

**Comment by AlexandreEichenberger** [comment]
> For future:
> 
> 1. call the pass "expression folding" or "expression canonicalization" ?

I can if you really want to... renaming files... but maybe better now than later

> 2. For the associativity, can we handle arbitrary movement? e.g.  (((v1+c1)+v2)+(v3+c2)) become (v1+(v2+(v3+(c1+c2)))) or similar one? Do we need introduce addOp with variant inputs for canonicalization?

I can get some... I can try a bit more aggressively. If I change (v1 + c) + v2 into (v1 + v2) + c systematically, constants would aggregate toward the root. An alternative would be to use a tensorflow like add with variadic ... but in presence of broadcast, ... let me try the simple case first.

> 3. For the result v1+c1 form, do we use PromoteAttribute?

I don't think we want that... promote attribute is good for small constants, I expect a lot of this arithmetic to be for weights... And we want them to be dumped in the binary using Tian's patch

**Pull request merged**